### PR TITLE
Added Halide backend support for deep learning layers

### PR DESCRIPTION
**Pull request by dkurt**
#### Implemented some layers using Halide language backend:

Layer | Covered all configurations | Test
---------------------------------|----------------------------|-----
Concat                           |          by channels (axis = 1)        | x
Convolution                      |                          x | x
Deconvolution                    |    no group factor    | x
(attachable) ReLU, TanH, Sigmoid, Abs, BNLL, Power, PReLU, Batch normalization, Scale | x | x
Element-wise addition, multiplication, maximum | x | x
Fully-connected | axis = 1 | x
LRN (across and within channels) | x | x
Maximum unpooling with indices support | x | x
Maximum pooling | x | x
Average pooling | only whole kernels, no padding | x
SoftMax | axis == 1, no spatial size more than 1x1 | x

Supported CPU only, float32 computations. No ahead-of-time compilation.
Added tutorial with installation guide and sample, tests for some layers and networks, performance tests.

#### Generalized manual scheduling of 5 networks into the layer-wise automatic scheduling.
Intel&reg; Core&trade; i7-6700K CPU @ 4.00GHz x 8

Architecture | Manual scheduling | Automatic scheduling | Ratio
------|------|------|------
AlexNet |            22.38ms |  25.5ms | x0.87
ResNet-50 | 73.91ms | 84.26ms | x0.87
SqueezeNet v1.1 | 8.21ms | 6.44ms | **x1.27**
Inception-5h | 37.06ms |  32.01ms | **x1.15**
ENet @ 3x512x256 | 41.21ms | 41.07ms| x1.00

For testing purpose have taken GoogLeNet network. L1 diff is 1.26789e-09. Efficiency is 33ms using Halide (automatic scheduling) against 34.89ms using MKL (x1.05).

TODO list:
- [x] Reproduce Halide installation by tutorial (Linux) https://github.com/halide/Halide/commit/abef1d9bf6cb3f866393fa4c5f48726f728285ee
- [ ] Reproduce Halide installation by tutorial (Windows): installation reproduced but SegFault appears (Some layers works with replacing vectorization by `x` variable to factor 4 (from 8 or 16). All right in case of no scheduling).
- [x] Cover more layers and configurations in tests
- [x] Write scheduling tutorial
- [x] Remove local paths
- [x] Figure out about `supportBackend` and layer configurations (axis, etc.)
- [x] Automatic scheduling with no more than 20% efficiency loss.

### CODE REVIEW

File: modules/dnn/tutorials/tutorial_dnn_halide.markdown [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+write image processing algorithms in well-readable format, schedule computations
+according to specific device and evaluate it with a quite good efficiency.
+
+An official website of the Halide project: http://halide-lang.org/.
+
+## Efficiency comparison
+Measured on Intel&reg; Core&trade; i7-6700K CPU @ 4.00GHz x 8.
+
+<html>
+  <table>
```
Comment (alalek) [inline modules/dnn/tutorials/tutorial_dnn_halide.markdown]: Please use markdown tables. [Example](https://github.com/opencv/opencv/blob/ec49eb813cbde55927d269a2e89a4dbb4a6b3262/modules/core/include/opencv2/core/hal/intrin_cpp.hpp#L183-L238). [Tutorial](https://help.github.com/articles/organizing-information-with-tables/)

File: modules/dnn/test/test_halide_nets.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+// Copyright (C) 2017, Intel Corporation, all rights reserved.
+// Third party copyrights are property of their respective owners.
+
+namespace cvtest
+{
+
+using namespace cv;
+using namespace dnn;
+
+void loadNet(const std::string& weights, const std::string& proto,
```
Comment (alalek) [inline modules/dnn/test/test_halide_nets.cpp]: please use "static" or "anonymous namespace" for internal functions.

BTW, Where is `#include` directive in this file?

File: modules/dnn/perf/perf_halide_net.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    net.setPreferableBackend(HALIDE); \
+    net.allocate(); \
+    net.applyHalideScheduler(scheduler); \
+    net.forward(net.getLayerId(outputLayer)); \
+    \
+    TEST_CYCLE_N(10) \
+    { \
+        net.forward(net.getLayerId(outputLayer)); \
+    } \
+    SANITY_CHECK_NOTHING();
```
Comment (dkurt) [inline modules/dnn/perf/perf_halide_net.cpp]: Thanks! Replaced to static function.

File: modules/dnn/perf/perf_halide_net.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+inline std::string getOpenCVExtraDir()
+{
+    return "/home/dkurtaev/opencv_extra/testdata/";
```
Comment (dkurt) [inline modules/dnn/perf/perf_halide_net.cpp]: Modified modules/dnn/perf/perf_main.cpp for it.

File: modules/dnn/CMakeLists.txt [hunk 5]
```diff
@@ -54,11 +54,34 @@ if(NOT PROTOBUF_FOUND)
...
+# ----------------------------------------------------------------------------
+if(HALIDE_FOUND)
+  set(DNN_INCLUDE_DIRS
+    ${DNN_INCLUDE_DIRS}
+    ${HALIDE_INCLUDE_DIRS}
+  )
+  set(DNN_LIBRARIES
+    ${DNN_LIBRARIES}
+    ${HALIDE_LIBRARIES}
+  )
```
Comment (dkurt) [inline modules/dnn/CMakeLists.txt]: Eliminated CMakeLists changes at all.

File: modules/dnn/tutorials/tutorial_dnn_halide.markdown (hunk 1 above)
Comment (dkurt) [inline modules/dnn/tutorials/tutorial_dnn_halide.markdown]: Fixed, thanks!

File: modules/dnn/test/test_halide_nets.cpp (hunk 2 above)
Comment (dkurt) [inline modules/dnn/test/test_halide_nets.cpp]: It's working somehow. Should I add it for prevent fails in different build cases?

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 6]
```diff
@@ -55,6 +55,23 @@ namespace dnn //! This namespace is used for dnn module functionlaity.
 
     typedef std::vector<int> MatShape;
 
+    /**
+     * @brief Enum of computation backends supported by layers.
+     */
+    enum Backend
+    {
+        DEFAULT,
```
Comment (vpisarev) [inline modules/dnn/include/opencv2/dnn/dnn.hpp]: I would rename them to DNN_BACKEND_DEFAULT, DNN_BACKEND_HALIDE, to make it more self-explanatory

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 7]
```diff
@@ -55,6 +55,23 @@ namespace dnn //! This namespace is used for dnn module functionlaity.
...
+        HALIDE
+    };
+
+    /**
+     * @brief Compilation targets of Halide code.
+     */
+    enum HalideTarget
+    {
+        HALIDE_CPU
```
Comment (vpisarev) [inline modules/dnn/include/opencv2/dnn/dnn.hpp]: again, let's use DNN_TARGET_CPU, DNN_TARGET_GPU etc. I would not use the word "Halide" here, because other backends, like OpenVX or even the default one, can be potentially customized to use CPU, GPU or other device to run the networks

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 8]
```diff
@@ -131,6 +148,47 @@ namespace dnn //! This namespace is used for dnn module functionlaity.
...
+        /**
+         * @brief Returns pointers to Halide functions.
+         * @param[in] inputs Pointers to input Halide buffers.
+         *
+         * Input buffers should be exactly the same that will be used in forward invocations.
+         * Despite we can use Halide::ImageParam based on input shape only,
+         * it helps prevent some memory management issues (if something wrong,
+         * Halide tests will be failed).
+         */
+        virtual std::vector<void*> initHalide(const std::vector<void*> &inputs);
```
Comment (vpisarev) [inline modules/dnn/include/opencv2/dnn/dnn.hpp]: instead of very low-level void* I would use some abstract class HalideFunc (that can encapsulate void*) with a smart pointer to it. It's comparatively low overhead, but it would be much more safe from memory management point of view.

Comment (dkurt) [inline modules/dnn/include/opencv2/dnn/dnn.hpp]: Introduced `BackendNode` and `BackendWrapper` whose derivatives store explicit `Halide::Func` and `Halide::Buffer` correspondingly. Moreover, it helps remove `Halide::Func` members from layers as it was before.

**Comment by vpisarev** [comment]
ok, let's merge it in while it's fresh! :+1:

**Comment by alalek** [comment]
This PR is rely on https://github.com/opencv/opencv/pull/8794 . Need to review/merge them too.
